quinn-proto/src/connection/assembler.rs

```diff
@@ -99,33 +107,49 @@ impl Assembler {
         let new = BinaryHeap::with_capacity(self.data.len());
         let old = mem::replace(&mut self.data, new);
         for chunk in old.into_sorted_vec().into_iter().rev() {
+            if !chunk.is_fragmented() {
+                self.data.push(chunk);
```

Comment:
Thought I added the comment, but apparently I messed something up: I'm not sure whether this handles the case where there is data from the previous fragment in `buffer`, which needs to be written before this. But then I'm also rather confused by the going through chunks in reversed order, and pushing here in forward order. 

Also let's keep in mind that data in `chunk` might overlap with other data, and we need to account for this (trim stuff, or make sure it at least stays in order).

Are all those paths currently covered by tests?

Reply:
Overlap can now never occur, because we call `deduplicate()` at the beginning of `defragment()`. The order of iteration is not really important, because `self.data` is a `BinaryHeap` that will keep the `Buffer` we push in order. However, iterating in that way will both enable the optimization of merging multiple `Buffer` into one and improve the performance of `self.data.push`.

Reply:
Oh, one thing that I got wrong was to assume that `self.data` is a vector or vecdeque, and that therefore you need to do `.push_front()` instead of `.push()`. However this doesn't apply, since this is a `BinaryHeap`.